File: capy/src/test/py/native_parser_smoke.py
import os
import importlib
from pathlib import Path

parser_provider = importlib.import_module("dev.capylang.compiler.parser.CapybaraParserProvider")


def source_kind(name):
    return {"__type": name}


def raw_module(name, path, source, kind):
    return {
        "__type": "RawModule",
        "name": name,
        "path": path,
        "input": source,
        "sourceKind": source_kind(kind),
    }


def assert_equal(actual, expected, label):
    if actual != expected:
        raise AssertionError(f"{label}: expected {expected!r}, got {actual!r}")


def generated_parser():
    parser = getattr(parser_provider, "parser", None)
    if parser is not None:
        return parser

    parser_functions = [
        getattr(parser_provider, name)
        for name in dir(parser_provider)
        if name.startswith("parser__") and callable(getattr(parser_provider, name))
    ]
    if len(parser_functions) != 1:
        raise AssertionError(f"expected one generated parser function, got {len(parser_functions)}")
    return parser_functions[0]


def main():
    parser_impl = generated_parser()().unsafe_run()

    parsed = parser_impl.parse(
        [
            raw_module(
                "ImportedFunctional",
                "dev/capylang/smoke",
                'from /capy/lang/String import { String }\n'
                'data User { name: String }\n'
                'fun greet(user: User): String = "hello {user.name}"\n',
                "FUNCTIONAL",
            ),
            raw_module(
                "Clock",
                "dev/capylang/smoke",
                "interface Clock {\n"
                "  def now(): int\n"
                "}\n"
                "class FixedClock(value: int): Clock {\n"
                '  field name: String = "fixed"\n'
                "  def now(): int = value\n"
                "}\n",
                "OBJECT_ORIENTED",
            ),
        ]
    )

    assert_equal(parsed["__type"], "ParsedProgram", "parsed type")
    assert_equal(len(parsed["modules"]), 2, "parsed module count")
    assert_equal(parsed["modules"][0]["imports"][0]["modulePath"], "/capy/lang/String", "import module")
    assert_equal(parsed["modules"][0]["definitions"][0]["__type"], "DataDeclaration", "first functional definition")
    assert_equal(parsed["modules"][1]["objectOriented"]["interfaces"][0]["name"], "Clock", "object interface")
    assert_equal(parsed["modules"][1]["objectOriented"]["classes"][0]["name"], "FixedClock", "object class")

    output = os.environ.get("PYTHON_NATIVE_PARSER_SMOKE_OUTPUT")
    if output:
        Path(output, "native-parser-smoke.txt").write_text("ok\n", encoding="utf-8")


if __name__ == "__main__":
    main()

File: capy/src/test/js/native_parser_smoke.js
'use strict';

const fs = require('fs');
const path = require('path');

const { parser } = require('dev/capylang/compiler/parser/CapybaraParserProvider.js');

function sourceKind(name) {
    return { __type: name };
}

function rawModule(name, modulePath, source, kind) {
    return {
        __type: 'RawModule',
        name,
        path: modulePath,
        input: source,
        sourceKind: sourceKind(kind),
    };
}

function unsafeRun(effect) {
    let value = effect;
    while (value && (typeof value.unsafe_run === 'function' || typeof value.unsafeRun === 'function')) {
        value = typeof value.unsafe_run === 'function' ? value.unsafe_run() : value.unsafeRun();
    }
    return value;
}

function assertEqual(actual, expected, label) {
    if (actual !== expected) {
        throw new Error(`${label}: expected ${JSON.stringify(expected)}, got ${JSON.stringify(actual)}`);
    }
}

const modules = [
    rawModule(
        'ImportedFunctional',
        'dev/capylang/smoke',
        'from /capy/lang/String import { String }\n'
            + 'data User { name: String }\n'
            + 'fun greet(user: User): String = "hello {user.name}"\n',
        'FUNCTIONAL'
    ),
    rawModule(
        'Clock',
        'dev/capylang/smoke',
        'interface Clock {\n'
            + '  def now(): int\n'
            + '}\n'
            + 'class FixedClock(value: int): Clock {\n'
            + '  field name: String = "fixed"\n'
            + '  def now(): int = value\n'
            + '}\n',
        'OBJECT_ORIENTED'
    ),
];

const parserImpl = unsafeRun(parser());
const parsed = parserImpl.parse(modules);

assertEqual(parsed.__type, 'ParsedProgram', 'parsed type');
assertEqual(parsed.modules.length, 2, 'parsed module count');
assertEqual(parsed.modules[0].imports[0].modulePath, '/capy/lang/String', 'import module');
assertEqual(parsed.modules[0].definitions[0].__type, 'DataDeclaration', 'first functional definition');
assertEqual(parsed.modules[1].objectOriented.interfaces[0].name, 'Clock', 'object interface');
assertEqual(parsed.modules[1].objectOriented.classes[0].name, 'FixedClock', 'object class');

try {
    const { compile } = require('dev/capylang/compiler/CapybaraCompiler.js');
    const { empty_native_provider_manifest } = require('dev/capylang/compiler/CompiledProgram.js');
    const emptyLibraries = { to_list: () => [] };
    const compilerResult = unsafeRun(compile(
        [
            rawModule('Simple', 'dev/capylang/smoke', 'fun one(): int = 1\n', 'FUNCTIONAL'),
            rawModule('SimpleObject', 'dev/capylang/smoke', 'interface SimpleObject {\n  def value(): int\n}\n', 'OBJECT_ORIENTED'),
        ],
        emptyLibraries,
        empty_native_provider_manifest(),
        empty_native_provider_manifest()
    ));
    assertEqual(compilerResult.__type, 'Left', 'compiler result');
} catch (error) {
    throw new Error(`generated JavaScript compiler smoke failed: ${error.message}`);
}

const output = process.env.JAVASCRIPT_NATIVE_PARSER_SMOKE_OUTPUT;
if (output) {
    fs.mkdirSync(output, { recursive: true });
    fs.writeFileSync(path.join(output, 'native-parser-smoke.txt'), 'ok\n', 'utf8');
}
